skimage/measure/mc_meta/visual_test.py

```diff
@@ -3,70 +3,118 @@
+def donuts():
+    """Return volume of two donuts based on a formula by Thomas Lewiner."""
     n = 48
-    a, b = 2.5 / n, -1.25
-    isovalue = 0.0
-    #
+    a = 2.5 / n * 8.0
+    b = -1.25 * 8.0
+    c = 16.0 - 1.85 * 1.85
+    d = 64.0
     vol = np.empty((n, n, n), 'float32')
-    for iz in range(vol.shape[0]):
-        for iy in range(vol.shape[1]):
-            for ix in range(vol.shape[2]):
-                z, y, x = float(iz)*a+b, float(iy)*a+b, float(ix)*a+b
-                vol[iz, iy, ix] = ( (
-                    (8*x)**2 + (8*y-2)**2 + (8*z)**2 + 16 - 1.85*1.85 ) * ( (8*x)**2 +
-                    (8*y-2)**2 + (8*z)**2 + 16 - 1.85*1.85 ) - 64 * ( (8*x)**2 + (8*y-2)**2 )
-                    ) * ( ( (8*x)**2 + ((8*y-2)+4)*((8*y-2)+4) + (8*z)**2 + 16 - 1.85*1.85 )
-                    * ( (8*x)**2 + ((8*y-2)+4)*((8*y-2)+4) + (8*z)**2 + 16 - 1.85*1.85 ) -
-                    64 * ( ((8*y-2)+4)*((8*y-2)+4) + (8*z)**2
-                    ) ) + 1025
-    # Uncommenting the line below will yield different results for classic MC
-    #vol = -vol
-
-elif SELECT == 4:
-    vol = ellipsoid(4, 3, 2, levelset=True)
-    isovalue = 0
-
-# Get surface meshes
-t0 = time.time()
-vertices1, faces1, _, _ = marching_cubes_lewiner(vol, isovalue, gradient_direction=gradient_dir, use_classic=False)
-print('finding surface lewiner took %1.0f ms' % (1000*(time.time()-t0)) )
-
-t0 = time.time()
-vertices2, faces2 = marching_cubes_classic(vol, isovalue, gradient_direction=gradient_dir)
-print('finding surface classic took %1.0f ms' % (1000*(time.time()-t0)) )
-
-# Show
-vv.figure(1); vv.clf()
-a1 = vv.subplot(121); m1 = vv.mesh(np.fliplr(vertices1), faces1)
-a2 = vv.subplot(122); m2 = vv.mesh(np.fliplr(vertices2), faces2)
-a1.camera = a2.camera
-
-# visvis uses right-hand rule, gradient_direction param uses left-hand rule
-m1.cullFaces = m2.cullFaces = 'front'  # None, front or back
-
-vv.use().Run()
+    for iz in range(n):
+        z = iz * a + b
+        z *= z
+        zc = z + c
+        for iy in range(n):
+            y = iy * a + b
+            y1 = y - 2
+            y1 *= y1
+            y2 = y + 2
+            y2 *= y2
+            for ix in range(n):
+                x = ix * a + b
+                x *= x
+                x1 = x + y1 + zc
+                x1 *= x1
+                x2 = x + y2 + zc
+                x2 *= x2
+                vol[iz, iy, ix] = ((x1 - d * (x + y1)) * (x2 - d * (z + y2)))
+    vol += 1025.0
+    return vol
+
+
+class Timer(object):
+    """Context manager for timing execution speed of body."""
+
+    def __init__(self, message):
+        """Initialize timer."""
+        print(message, end=' ')
+        self.start = 0
+
+    def __enter__(self):
+        """Enter timer context."""
+        self.start = time.time()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Exit timer context."""
+        print('took %1.0f ms' % (1000 * (time.time() - self.start)))
+

```

Comment:
This is fine, but imho it is more readable using the contextlib contextmanager decorator:

```python
from contextlib import contextmanager

@contextmanager
def timer(message):
    print(message, end=' ')
    start = time.time()
    yield
    print('took {:1.0f} ms'.format(1000 * (time.time() - start)))
```

Reply:
@jni, is there a way to make context managers for devices? 

In my own projects, i have "classes" to describe hardware devices, and I would like to add context managers that would simply call the "open" and "close" functions for me. Ideas?

Reply:
Why not add `__enter__` and `__exit__` methods to the "classes"?

Reply:
yeah, that is what I do now. Wondering if that is the correct approach or not.